
from dataclasses import dataclass
import typing
import typing_extensions

from flytekit.core.annotation import FlyteAnnotation

from latch.types.metadata import NextflowParameter
from latch.types.file import LatchFile
from latch.types.directory import LatchDir, LatchOutputDir

# Import these into your `__init__.py` file:
#
# from .parameters import generated_parameters

generated_parameters = {
    'input': NextflowParameter(
        type=typing.Optional[LatchFile],
        default=None,
        section_title='Input/output options',
        description='Path to comma-separated file containing information about the samples in the experiment.',
    ),
    'outdir': NextflowParameter(
        type=typing_extensions.Annotated[LatchDir, FlyteAnnotation({'output': True})],
        default=None,
        section_title=None,
        description='The output directory where the results will be saved. You have to use absolute paths to storage on Cloud infrastructure.',
    ),
    'bait_padding': NextflowParameter(
        type=typing.Optional[int],
        default=100,
        section_title='Reference file options',
        description='The amount to pad each end of the target intervals to create bait intervals.',
    ),
    'bwa': NextflowParameter(
        type=typing.Optional[LatchDir],
        default=None,
        section_title=None,
        description='Directory for pre-built bwa index.',
    ),
    'bwamem2': NextflowParameter(
        type=typing.Optional[LatchDir],
        default=None,
        section_title=None,
        description='Directory for pre-built bwamem2 index.',
    ),
    'bwameme': NextflowParameter(
        type=typing.Optional[LatchDir],
        default=None,
        section_title=None,
        description="Directory for pre-built bwameme's learned index.",
    ),
    'cadd_resources': NextflowParameter(
        type=typing.Optional[LatchDir],
        default=None,
        section_title=None,
        description='Path to the directory containing cadd annotations.',
    ),
    'fai': NextflowParameter(
        type=typing.Optional[LatchFile],
        default=None,
        section_title=None,
        description='Path to FASTA genome index file.',
    ),
    'fasta': NextflowParameter(
        type=LatchFile,
        default=None,
        section_title=None,
        description='Path to FASTA genome file.',
    ),
    'gcnvcaller_model': NextflowParameter(
        type=typing.Optional[LatchFile],
        default=None,
        section_title=None,
        description='A file containing the path to models produced by GATK4 GermlineCNVCaller cohort.',
    ),
    'genome': NextflowParameter(
        type=typing.Optional[str],
        default='GRCh38',
        section_title=None,
        description='Name of iGenomes reference.',
    ),
    'gnomad_af': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Path to the gnomad tab file with allele frequencies.',
    ),
    'gnomad_af_idx': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Path to the index file for the gnomad tab file with allele frequencies.',
    ),
    'igenomes_ignore': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Do not load the iGenomes reference config.',
    ),
    'intervals_wgs': NextflowParameter(
        type=str,
        default=None,
        section_title=None,
        description='Path to the interval list of the genome (autosomes, sex chromosomes, and mitochondria).',
    ),
    'intervals_y': NextflowParameter(
        type=str,
        default=None,
        section_title=None,
        description='Path to the interval list of the Y chromosome.',
    ),
    'known_dbsnp': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Path to known dbSNP file.',
    ),
    'known_dbsnp_tbi': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Path to known dbSNP file index.',
    ),
    'local_genomes': NextflowParameter(
        type=typing.Optional[LatchDir],
        default=None,
        section_title=None,
        description='Local directory base for genome references that map to the config.',
    ),
    'mito_name': NextflowParameter(
        type=typing.Optional[str],
        default='chrM',
        section_title=None,
        description='Name of the mitochondrial contig in the reference fasta file',
    ),
    'mobile_element_references': NextflowParameter(
        type=typing.Optional[LatchFile],
        default=None,
        section_title=None,
        description='File with mobile element references',
    ),
    'mobile_element_svdb_annotations': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='File with mobile element allele frequency references',
    ),
    'ml_model': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Path to sentieon machine learning model file.',
    ),
    'mt_fasta': NextflowParameter(
        type=typing.Optional[LatchFile],
        default=None,
        section_title=None,
        description='Path to mitochondrial FASTA genome file.',
    ),
    'ploidy_model': NextflowParameter(
        type=typing.Optional[LatchDir],
        default=None,
        section_title=None,
        description='Directory containing the ploidy model files',
    ),
    'readcount_intervals': NextflowParameter(
        type=typing.Optional[LatchFile],
        default=None,
        section_title=None,
        description='Interval list file containing the intervals over which read counts are tabulated for CNV calling',
    ),
    'reduced_penetrance': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='File with gene ids that have reduced penetrance. For use with genmod',
    ),
    'rtg_truthvcfs': NextflowParameter(
        type=typing.Optional[LatchFile],
        default=None,
        section_title=None,
        description='Vcf used for evaluating variant calls.',
    ),
    'save_reference': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='If generated by the pipeline save the required indices/references in the results directory.',
    ),
    'score_config_mt': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='MT rank model config file for genmod.',
    ),
    'score_config_snv': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='SNV rank model config file for genmod.',
    ),
    'score_config_sv': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='SV rank model config file for genmod.',
    ),
    'sdf': NextflowParameter(
        type=typing.Optional[LatchDir],
        default=None,
        section_title=None,
        description='Directory for pre-built sdf index. Used by rtg/vcfeval',
    ),
    'sequence_dictionary': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Path to the genome dictionary file',
    ),
    'svdb_query_bedpedbs': NextflowParameter(
        type=typing.Optional[LatchFile],
        default=None,
        section_title=None,
        description='Databases used for structural variant annotation in chrA-posA-chrB-posB-type-count-frequency format.',
    ),
    'svdb_query_dbs': NextflowParameter(
        type=typing.Optional[LatchFile],
        default=None,
        section_title=None,
        description='Databases used for structural variant annotation in vcf format.',
    ),
    'target_bed': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Path to directory for target bed file.',
    ),
    'variant_catalog': NextflowParameter(
        type=typing.Optional[LatchFile],
        default=None,
        section_title=None,
        description='Path to variant catalog file',
    ),
    'sample_id_map': NextflowParameter(
        type=typing.Optional[LatchFile],
        default=None,
        section_title=None,
        description='Path to a file containing internal ids and customer ids in csv format.',
    ),
    'vcf2cytosure_blacklist': NextflowParameter(
        type=typing.Optional[LatchFile],
        default=None,
        section_title=None,
        description='Path to vcf2cytosure blacklist file',
    ),
    'vcfanno_resources': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Path to a file containing the absolute paths to resources defined within the vcfanno toml file. One line per resource.',
    ),
    'vcfanno_toml': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Path to the vcfanno toml file.',
    ),
    'vcfanno_lua': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Path to the vcfanno lua file.',
    ),
    'vep_cache': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description="Path to vep's cache directory.",
    ),
    'vep_plugin_files': NextflowParameter(
        type=typing.Optional[LatchFile],
        default=None,
        section_title=None,
        description='Databases used by both named and custom plugins to annotate variants.',
    ),
    'vep_filters': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Path to the file containing HGNC_IDs of interest on separate lines.',
    ),
    'vep_filters_scout_fmt': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='Path to a bed-like file exported by scout, which contains HGNC_IDs to be used in filter_vep.',
    ),
    'analysis_type': NextflowParameter(
        type=typing.Optional[str],
        default='wgs',
        section_title='Analysis options',
        description="Specifies which analysis type for the pipeline- either 'wgs','wes','mito'. This changes resources consumed and tools used.",
    ),
    'bwa_as_fallback': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Specifies whether or not to use bwa as a fallback aligner in case bwamem2 throws an error.',
    ),
    'platform': NextflowParameter(
        type=typing.Optional[str],
        default='illumina',
        section_title=None,
        description="Specifies which analysis type for the pipeline- either 'wgs','wes','mito'. This changes resources consumed and tools used.",
    ),
    'ngsbits_samplegender_method': NextflowParameter(
        type=typing.Optional[str],
        default='xy',
        section_title=None,
        description='Method selection for ngs-bits samplegender',
    ),
    'run_mt_for_wes': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Specifies whether to run mitochondrial analysis for wes samples',
    ),
    'run_rtgvcfeval': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description="Specifies whether to run rtgtools' vcfeval",
    ),
    'save_mapped_as_cram': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Specifies whether to generate and publish alignment files as cram instead of bam',
    ),
    'skip_fastqc': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Specifies whether or not to skip FASTQC.',
    ),
    'skip_fastp': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Specifies whether or not to skip trimming with fastp.',
    ),
    'skip_haplocheck': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Specifies whether or not to skip haplocheck.',
    ),
    'skip_gens': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Specifies whether or not to skip gens preprocessing subworkflow.',
    ),
    'skip_germlinecnvcaller': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description="Specifies whether or not to skip CNV calling using GATK's GermlineCNVCaller",
    ),
    'skip_eklipse': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Specifies whether or not to skip eKLIPse.',
    ),
    'skip_peddy': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Specifies whether or not to skip peddy.',
    ),
    'skip_qualimap': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Specifies whether or not to skip Qualimap.',
    ),
    'skip_me_calling': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Specifies whether or not to skip calling mobile elements, and the subsequent annotation step.',
    ),
    'skip_me_annotation': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Specifies whether or not to skip annotation of mobile elements.',
    ),
    'skip_mt_annotation': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Specifies whether or not to skip annotation of mitochondrial variants.',
    ),
    'skip_mt_subsample': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Specifies whether or not to subsample mt alignment.',
    ),
    'skip_snv_annotation': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Specifies whether or not to skip annotate SNV subworkflow.',
    ),
    'skip_sv_annotation': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Specifies whether or not to skip annotate structural variant subworkflow.',
    ),
    'skip_vcf2cytosure': NextflowParameter(
        type=typing.Optional[bool],
        default=True,
        section_title=None,
        description='Specifies whether or not to skip the vcf2cytosure subworkflow',
    ),
    'skip_vep_filter': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description="Specifies whether or not to filter results based on a list of candidate genes specified in 'vep_filters'.",
    ),
    'aligner': NextflowParameter(
        type=typing.Optional[str],
        default='bwamem2',
        section_title='Alignment options',
        description="Specifies the alignment algorithm to use - available options are 'bwamem2' and 'sentieon'.",
    ),
    'min_trimmed_length': NextflowParameter(
        type=typing.Optional[int],
        default=40,
        section_title=None,
        description='Discard trimmed reads shorter than the given value',
    ),
    'mt_subsample_rd': NextflowParameter(
        type=typing.Optional[int],
        default=150,
        section_title=None,
        description='Expected coverage to subsample mt alignment to.',
    ),
    'mt_subsample_seed': NextflowParameter(
        type=typing.Optional[int],
        default=30,
        section_title=None,
        description='Subsampling seed used to influence which subset of mitochondrial reads is kept. ',
    ),
    'rmdup': NextflowParameter(
        type=typing.Optional[bool],
        default=None,
        section_title=None,
        description='Specifies whether duplicates reads should be removed prior to variant calling.',
    ),
    'call_interval': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title='Variant calling options',
        description='Interval in the reference that will be used in the software',
    ),
    'cnvnator_binsize': NextflowParameter(
        type=typing.Optional[int],
        default=1000,
        section_title=None,
        description='Bin size for CNVnator',
    ),
    'sentieon_dnascope_pcr_indel_model': NextflowParameter(
        type=typing.Optional[str],
        default='CONSERVATIVE',
        section_title=None,
        description='Option for selecting the PCR indel model used by Sentieon Dnascope.',
    ),
    'variant_caller': NextflowParameter(
        type=typing.Optional[str],
        default='deepvariant',
        section_title=None,
        description="Specifies the variant caller to use - available options are 'deepvariant' and 'sentieon'.",
    ),
    'variant_type': NextflowParameter(
        type=typing.Optional[str],
        default='snp,indel',
        section_title=None,
        description='Specifies the variant types for sentieon variant caller.',
    ),
    'variant_consequences_snv': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title='Annotation options',
        description='File containing list of SO terms listed in the order of severity from most severe to lease severe for annotating genomic and mitochondrial SNVs.',
    ),
    'variant_consequences_sv': NextflowParameter(
        type=typing.Optional[str],
        default=None,
        section_title=None,
        description='File containing list of SO terms listed in the order of severity from most severe to lease severe for annotating genomic SVs.',
    ),
    'vep_cache_version': NextflowParameter(
        type=typing.Optional[int],
        default=110,
        section_title=None,
        description='Specify the version of the VEP cache provided to the `--vep_cache` option.',
    ),
}

